From d4e984a372eb96f0fcbd683556ab5986ef7faa3f Mon Sep 17 00:00:00 2001
From: Michal Simek <monstr@monstr.eu>
Date: Thu, 26 Apr 2012 07:50:33 +0200
Subject: [PATCH] zynq: smp: Do not ioremap the jump trampoline

If CONFIG_PHYS_OFFSET is setup to zero, physical ram
address 0x0 is already mapped that's why not possible
to map it again. This is valid only for RAM not of IO.

This should be the part of patch:
zynq: Fix boot_secondary to directly reset cpu1
(sha1: 61746d965796f88e3bfc49da2267cea006471269)

Log:
Calibrating local timer... 399.27MHz.
------------[ cut here ]------------
WARNING: at arch/arm/mm/ioremap.c:241 __arm_ioremap_pfn_caller+0x108/0x1a4()
Modules linked in:
[<c0012c10>] (unwind_backtrace+0x0/0xe0) from [<c001ab64>] (warn_slowpath_common+0x4c/0x64)
[<c001ab64>] (warn_slowpath_common+0x4c/0x64) from [<c001ab94>] (warn_slowpath_null+0x18/0x1c)
[<c001ab94>] (warn_slowpath_null+0x18/0x1c) from [<c0014ca4>] (__arm_ioremap_pfn_caller+0x108/0x1a4)
[<c0014ca4>] (__arm_ioremap_pfn_caller+0x108/0x1a4) from [<c0014da0>] (__arm_ioremap_caller+0x50/0x54)
[<c0014da0>] (__arm_ioremap_caller+0x50/0x54) from [<c048bb9c>] (platform_smp_prepare_cpus+0x18/0x68)
[<c048bb9c>] (platform_smp_prepare_cpus+0x18/0x68) from [<c04867fc>] (kernel_init+0x44/0x130)
[<c04867fc>] (kernel_init+0x44/0x130) from [<c000e3c0>] (kernel_thread_exit+0x0/0x8)
---[ end trace 1b75b31a2719ed1c ]---
!!!! BOOTUP jump vectors can't be used !!!!

Signed-off-by: Michal Simek <monstr@monstr.eu>
---
 arch/arm/mach-zynq/platsmp.c |   10 +++++++++-
 1 files changed, 9 insertions(+), 1 deletions(-)

diff --git a/arch/arm/mach-zynq/platsmp.c b/arch/arm/mach-zynq/platsmp.c
index 7f3ae25..2d942fd 100644
--- a/arch/arm/mach-zynq/platsmp.c
+++ b/arch/arm/mach-zynq/platsmp.c
@@ -169,10 +169,18 @@ void __init platform_smp_prepare_cpus(unsigned int max_cpus)
 	 * Remap the first three addresses at zero which are used
 	 * for 32bit long jump for SMP. Look at zynq_cpu1_start()
 	 */
+#if CONFIG_PHYS_OFFSET
 	zero = ioremap(0, 12);
-	if (!zero)
+	if (!zero) {
 		printk(KERN_WARNING
 			"!!!! BOOTUP jump vectors can't be used !!!!\n");
+		while (1)
+			;
+	}
+#else
+	/* The first three addresses at zero are already mapped */
+	zero = (u8 *)CONFIG_PAGE_OFFSET;
+#endif
 
 	/*
 	 * Initialise the present map, which describes the set of CPUs
-- 
1.7.0.4

